Add Ploi

Cannonb4ll opened the pull request:
## Description

This is an extension for [ploi.io](ploi.io) where you can easily manage your servers & sites through Raycast.

There are quite a few features it has:

- View site details
- View deployment status
- Trigger deploy script
- Restart services (NGINX, MySQL, Supervisor)
- Restart server
- Flush FastCGI cache
- Refresh server OPcache

Original repository: https://github.com/ploi-deploy/ploi-raycast

![PloiRaycast](https://github.com/ploi-deploy/ploi-raycast/blob/main/screenshots/banner.png?raw=true)

## Type of change

- New extension

## Screencast

https://screenshot.webbuilds.nl/E6B4MLcBqF72rban4ZfV

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)

## Helpful links:

- https://developers.ploi.io
- https://ploi.io
- https://github.com/ploi-deploy/ploi-raycast

KevinBatdorf commented:
Do I get to be listed as a contributor? 😎 how did you even find my code? I only submitted it last night. 

One improvement I made recently was when triggering something like a deployment, I will fire a Toast message. You can use `new Toast()` then the `show` and `hide` methods. Otherwise the feedback felt less snappy. 

Was there anything notable you found that I could improve upon?

Cannonb4ll commented:
> Do I get to be listed as a contributor? 😎 how did you even find my code? I only submitted it last night.
> 
> One improvement I made recently was when triggering something like a deployment, I will fire a Toast message. You can use `new Toast()` then the `show` and `hide` methods. Otherwise the feedback felt less snappy.
> 
> Was there anything notable you found that I could improve upon?

I've outsourced this, and then made significant improvements on my side. I do see similarities. I will add you as a source of inspiration!

The one biggest thing I could find that was notably annoying was the constant rewrite of the headers part. So I made that as a default, you might want to consider that as well:

https://github.com/raycast/extensions/blob/52b0ee56bd7667a1aef3753a022286b2f0ea883c/extensions/ploi/src/index.tsx

This allows you to remove the headers part for each call, which was repetitive. Other than that I did some small improvements and changes to make the code a bit cleaner and in the style of Ploi (Some things are a bit different).

Feel free to take a look around and grab some code if you like! I think we can benefit from both of these in regards to improvements 🤝

Good call about the toast, by the way, I haven't looked at it much. One thing I was planning on adding the ability to add a new site to the server.

KevinBatdorf commented:
No need to credit anything I was just surprised how fast it came through. They must have saw my tweet from last week or something. Of course I have no issue with it though. 

I'll check out your suggestions. Thanks! Since the services are similar I'll look out for the site feature as well. I was hesitant to add that but I'll see how you impliment it. 

Have you considered tracking API usage that originates from the extension? I wouldn't expect you to share the data but it would be interesting I bet. I could infer the usage based on the amount features you invest here, I suppose! Maybe I'll ping Taylor about this too.

Cannonb4ll commented:
> No need to credit anything I was just surprised how fast it came through. They must have saw my tweet from last week or something. Of course I have no issue with it though.

I don't know if they did copy, what I do know is that I put out a bounty in our Ploi Discord community on the 13th of December 2021, that's a bit more over a week ago (https://screenshot.webbuilds.nl/Z3rO4EpCYoaLZBgGoCfv). This bounty was also picked up the same day as well. (Oh, and I did credit you in our original repository https://github.com/ploi-deploy/ploi-raycast#contributors 😊)

> I'll check out your suggestions. Thanks! Since the services are similar I'll look out for the site feature as well. I was hesitant to add that but I'll see how you impliment it.

Yeah, I think we can benefit both from the features we add!

> Have you considered tracking API usage that originates from the extension? I wouldn't expect you to share the data but it would be interesting I bet. I could infer the usage based on the amount features you invest here, I suppose! Maybe I'll ping Taylor about this too.

I have! I already made preparations for it: https://github.com/ploi-deploy/ploi-raycast/blob/main/src/index.tsx#L13, because we're going to use this header to also display in the server logs in ploi itself, you'll be able to see whether you did an action via Raycast, which is awesome. I have no data for this plugin yet as it has not been added as of yet, but I wouldn't mind sharing if it gets used a lot in the near future if this does get approved.

KevinBatdorf commented:
They left in a lot of the comments, quirks, ts-ignore bits though that I think make it clear. The whole approach and code structure and object naming is pretty identical as well. But copying is allowed (and smart if they got paid!) and I genuinely don't mind. I encourage it even. I was kidding at the start but maybe that was inappropriate of me to joke about.

Maybe someone can use it for a cloudways extension as well.

pernielsentikaer reviewed (changes requested):
Hi @Cannonb4ll 👋 

Seems like an awesome extension 🔥 

First, before I dig in, I'd like you to work on a solution that gives some feedback if an API key is wrong when using the command.

Right now it just loads forever without any toast telling there is an error (and stopping the loader with empty array (Then it shows “No Results”))

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

Cannonb4ll commented:
Hi @pernielsentikaer,

I've already pushed adjustments, that should make it a lot clearer. As of now, it will show a correct toast the API key is wrong, it will then pop back to the root view (as it will not display any results anyway).

If you want a working API key, please email me at dennis@ploi.io (as I don't want to expose tokens here), I'll get you sorted so you can test further. (This testing key can be used for random deploys etc. it doesn't matter).

pernielsentikaer reviewed (changes requested):
Hi @Cannonb4ll 👋 

Sorry, just checked my settings – I forgot that I created a new empty Ploi account and added the key, then it just keeps “loading” forever – here it should show “No Results” I guess?

I will e-mail you, so we can discuss further 👍

Cannonb4ll commented:
@pernielsentikaer I have just replied to your email, with a API key that you can use to test further, and also did 2 commits. 

- The toast for a invalid API key is now showing properly and will pop raycast back to root.
- When entering a valid API key without any servers it will just show an empty list item.

Let me know!

pernielsentikaer commented on extensions/ploi/src/Server.tsx:
```diff
@@ -0,0 +1,299 @@
+        content={server.ipAddress}
+      />
+      <OpenInBrowserAction
+        title="Open In ploi.io"
```
```suggestion
        title="Open in ploi.io"
```

pernielsentikaer commented on extensions/ploi/src/Site.tsx:
```diff
@@ -0,0 +1,224 @@
+          <List.Item
+            id="open-in-ploi"
+            key="open-in-ploi"
+            title="Open In ploi.io"
```
```suggestion
            title="Open in ploi.io"
```

pernielsentikaer reviewed (changes requested):
Hi @Cannonb4ll 👋 

Much better, I left some feedback and suggestions for you 😄 

- Instead of using `eslint-disable-next-line` with `@ts-ignore`, then you can use `@ts-expect-error` instead (which is accepted by eslint as long as you explain why there is an error, as you already did) – That's a bit cleaner

- `Open Server Info` is missing an icon.

- When using the `Open Server Info` I just get this:

![hc5aXEbY7AEAZpvE](https://user-images.githubusercontent.com/38033107/149275612-cc42bba1-1fb4-4e71-8459-e966bd817373.png)

```
2022-01-13 07:13:04.229 F  Raycast[39916:1db072] [com.raycast.extensions.ploi:index] /Users/pernielsentikaer/Desktop/2021-12-06/2021-12-06/rc2/extensions/ploi/src/helpers.ts:35
    pollingId.current = window.setInterval(poll, 60000 / API_RATE_LIMIT + 500);
                        ^

ReferenceError: window is not defined
    at null.<anonymous> (/Users/pernielsentikaer/Desktop/2021-12-06/2021-12-06/rc2/extensions/ploi/src/helpers.ts:35:25)
    at bo (/Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:9:4879)
    at C.unstable_runWithPriority (/Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:1:5006)
    at Ke (/Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:5:4136)
    at be (/Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:9:4348)
    at /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:9:4262
    at du (/Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:1:4027)
    at Timeout.uu [as _onTimeout] (/Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/react-reconciler/react-reconciler.production.min.js:1:1409)
    at listOnTimeout (node:internal/timers:557:17)
    at processTimers (node:internal/timers:500:7)
2022-01-13 07:13:04.242 F  Raycast[39916:1db072] [com.raycast.extensions.ploi:index]
ReferenceError: window is not defined

SitesList:Site.tsx:17:3

---
14: import { PLOI_PANEL_URL } from "./config";
15:
16: export const SitesList = ({
17:   server,
18:   sites: sitesArray,
19: }: {
20:   server: IServer;
---

    at ray-list-section
ta:index.js:25:38388
    at ray-list
ci:index.js:25:37716
SingleServerView:Server.tsx:126:3
ei:index.js:25:22505
    at ray-navigation-stack
od:index.js:25:22138
qo:index.js:25:22243
ni:index.js:25:22843
$A:index.js:25:20707
    at ray-root
wu:index.js:25:23147
```
- When opening the command it flickers `If you create your first server in ploi.io, it will show up here.` Is there a way to only show this if needed?

Looking forward for the next update already 🔥 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

Cannonb4ll commented:
> Instead of using eslint-disable-next-line with @ts-ignore, then you can use @ts-expect-error instead (which is accepted by eslint as long as you explain why there is an error, as you already did) – That's a bit cleaner

Done ✅

> Open Server Info is missing an icon.

I'm not sure about this one, I was not able to find a missing icon. Can you elaborate?

> When using the Open Server Info I just get this:

Fixed ✅

> When opening the command it flickers If you create your first server in ploi.io, it will show up here. Is there a way to only show this if needed?

Fixed ✅

Ready for a new review.

pernielsentikaer reviewed (changes requested):
Hi @Cannonb4ll 👋 

Thanks for the update, what an awesome extension, thanks for the demo API key, it makes it alot more easy to test 💯 

The icons missing is this two places:

![EWmJ1EFdhcUqFcDw](https://user-images.githubusercontent.com/38033107/149418262-bf76da88-fc05-4f06-9f5c-0dfc4de8215e.png)

![JMPTm3N2Whakc104](https://user-images.githubusercontent.com/38033107/149418285-d641f101-5fd3-4e12-99e1-2a12cdc9ebe9.png)



There is also no way to uppercase every accessoryTitle word like here:

![image](https://user-images.githubusercontent.com/38033107/149418131-87fd23a3-028a-40eb-bfac-9026f4e5906f.png)
I think it will look better without actually

Could you look into this?

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

pernielsentikaer reviewed (approved):
Hi @Cannonb4ll 👋 

Awesome! Look so much better 🔥 I'm going to accept it and then will my colleague do a final review before it's finally merged.
